File: LINKED LISTS/Delete Operations/DeleteAtStart.java
public class DeleteAtStart {

    // ======================================================
    // NODE CLASS
    // ======================================================

    static class Node {
        int data;
        Node next;
    }


    // ======================================================
    // CREATE A NEW NODE
    // ======================================================

    static Node createNode(int data) {
        Node newNode = new Node();
        newNode.data = data;
        newNode.next = null;
        return newNode;
    }


    // ======================================================
    // TRAVERSE THE LIST
    // ======================================================

    static void traverse(Node head) {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }


    // ======================================================
    // DELETE AT THE BEGINNING
    // ======================================================

    /*
        Core pointer idea:

            head = head.next;

        That single line IS delete-at-start in Java.
        The old first node simply loses its last reference
        and the garbage collector reclaims it.

        In C this needed three steps because YOU are the
        memory manager:

            struct Node *temp = head;   // 1. save the address
            head = head->next;          // 2. move head forward
            free(temp);                 // 3. release [10]

        In Java, steps 1 and 3 disappear — there is no free()
        and no way to "lose" an address we still need. Once no
        reference points to the old node, it is unreachable,
        and unreachable objects are collected automatically.
    */

    static Node deleteAtBeginning(Node head) {

        /*
            CASE 1: The list is empty.

                head
                 ↓
                null

            Nothing to delete. Return head unchanged.
        */

        if (head == null) {
            System.out.println("List is empty. Nothing to delete.");
            return head;
        }


        /*
            CASE 2: The list has one or more nodes.

                head
                 ↓
               [10] → [20] → [30] → null

            We want to remove [10] and make [20] the new head.

                head = head.next;

            State after:

                     head
                      ↓
               [10] → [20] → [30] → null

            [10] is now orphaned — nothing points to it anymore.
            In C we would have had to free(temp) at this point.
            In Java, the garbage collector will clean up [10]
            on its own. We never touch it again.
        */

        head = head.next;   // advance head; the old node becomes garbage

        return head;        // return new head so caller can update their variable
    }


    // ======================================================
    // MAIN
    // ======================================================

    public static void main(String[] args) {

        /*
            Build the list:  10 → 20 → 30 → null
        */

        Node head = createNode(10);
        head.next = createNode(20);
        head.next.next = createNode(30);

        System.out.print("Original list:             ");
        traverse(head);
        // 10 20 30

        /*
            Delete the first node.

            Before:   10 → 20 → 30
                       ↓
            After:         20 → 30
        */

        head = deleteAtBeginning(head);

        System.out.print("After deleting at start:   ");
        traverse(head);
        // 20 30

        /*
            Delete again — now 20 is the first node.

            Before:   20 → 30
                       ↓
            After:         30
        */

        head = deleteAtBeginning(head);

        System.out.print("After deleting at start:   ");
        traverse(head);
        // 30

        /*
            Delete the last remaining node.

            Before:   30
                       ↓
            After:    (empty list)
        */

        head = deleteAtBeginning(head);

        System.out.print("After deleting at start:   ");
        traverse(head);
        // (blank line — empty list)

        /*
            Try to delete from an empty list.
            Should print a message and do nothing.
        */

        head = deleteAtBeginning(head);

        /*
            No need to free memory in Java.
            The garbage collector cleans up automatically
            once the deleted nodes become unreachable.
        */
    }
}

File: LINKED LISTS/Delete Operations/DeleteAtStart.c
#include <stdio.h>
#include <stdlib.h>


// ==============================
// NODE STRUCTURE
// ==============================

struct Node {
    int data;
    struct Node *next;
};


// ==============================
// HELPER: CREATE A NEW NODE
// ==============================

struct Node *createNode(int data) {
    struct Node *newNode = malloc(sizeof(struct Node));
    if (newNode == NULL) return NULL;
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}


// ==============================
// HELPER: TRAVERSE THE LIST
// ==============================

void traverse(struct Node *head) {
    struct Node *current = head;
    while (current != NULL) {
        printf("%d ", current->data);
        current = current->next;
    }
    printf("\n");
}


// ==============================
// DELETE AT THE BEGINNING
// ==============================

struct Node *deleteAtBeginning(struct Node *head) {

    /*
        CASE 1: The list is empty.

        head
         ↓
        NULL

        There is nothing to delete. Return head unchanged.
    */

    if (head == NULL) {
        printf("List is empty. Nothing to delete.\n");
        return head;
    }


    /*
        CASE 2: The list has one or more nodes.

        head
         ↓
        [10] → [20] → [30] → NULL

        We want to remove [10] and make [20] the new head.

        STEP 1: Save the current head in a temporary pointer.

            struct Node *temp = head;

        State:

            temp         head
              ↓            ↓
             [10] → [20] → [30] → NULL

            Wait — both temp and head point to [10].
            We need temp so we can free [10] after we
            move head forward.


        STEP 2: Advance head to the next node.

            head = head->next;

        State:

            temp    head
              ↓       ↓
             [10] → [20] → [30] → NULL

        Now head correctly points to [20].
        But [10] is still sitting in memory — we must free it.


        STEP 3: Free the old head node.

            free(temp);

        State:

            head
              ↓
             [20] → [30] → NULL

        [10] is gone. [20] is the new head.


        Why do we need temp?

        If we did head = head->next first and then tried to
        free the old head, we would have already lost the
        address of [10]. We'd have no way to free it.

        temp preserves that address so free() can use it.
    */

    struct Node *temp = head;      // STEP 1: save the address of [10]
    head = head->next;             // STEP 2: move head forward to [20]
    free(temp);                    // STEP 3: free the old [10] node

    return head;                   // return new head so caller can update their pointer
}


// ==============================
// MAIN
// ==============================

int main() {

    /*
        Build the list:  10 → 20 → 30 → NULL
    */

    struct Node *head = createNode(10);
    head->next = createNode(20);
    head->next->next = createNode(30);

    printf("Original list:             ");
    traverse(head);
    // 10 20 30

    /*
        Delete the first node.

        Before:   10 → 20 → 30
                   ↓
        After:    20 → 30
    */

    head = deleteAtBeginning(head);

    printf("After deleting at start:   ");
    traverse(head);
    // 20 30

    /*
        Delete again — now 20 is the first node.

        Before:   20 → 30
                   ↓
        After:    30
    */

    head = deleteAtBeginning(head);

    printf("After deleting at start:   ");
    traverse(head);
    // 30

    /*
        Delete the last remaining node.

        Before:   30
                   ↓
        After:    (empty list)
    */

    head = deleteAtBeginning(head);

    printf("After deleting at start:   ");
    traverse(head);
    // (blank line — empty list)

    /*
        Try to delete from an empty list.
        Should print a message and do nothing.
    */

    head = deleteAtBeginning(head);

    return 0;
}
